fdbserver/KeyValueStoreRocksDB.actor.cpp

```diff
@@ -299,55 +303,128 @@ struct RocksDBKeyValueStore : IKeyValueStore {
 		a->path = path;
 		auto res = a->done.getFuture();
 		writeThread->post(a.release());
-		return res;
+		return join(res, sqlLite->init());
 	}
 
-	void set(KeyValueRef kv, const Arena*) override {
+	void set(KeyValueRef kv, const Arena* a) override {
+		// Writes to the special keyspace only go to SQLite.
+		if (kv.key >= SPECIAL_KEYSPACE) {
+			sqlLite->set(kv, a);
+			return;
+		}
 		if (writeBatch == nullptr) {
 			writeBatch.reset(new rocksdb::WriteBatch());
 		}
 		writeBatch->Put(toSlice(kv.key), toSlice(kv.value));
 	}
 
-	void clear(KeyRangeRef keyRange, const Arena*) override {
+	void clear(KeyRangeRef keyRange, const Arena* a) override {
+		// If the clear touches the special keyspace send the clear to SQLite.
+		if (keyRange.end > SPECIAL_KEYSPACE) {
+			sqlLite->clear(keyRange, a);
+		}
+		// If the clear doesn't touch user ranges at all, don't create extra tombstones in Rocks.
+		if (keyRange.begin >= SPECIAL_KEYSPACE) {
+			return;
+		}
+
 		if (writeBatch == nullptr) {
 			writeBatch.reset(new rocksdb::WriteBatch());
 		}
 
 		writeBatch->DeleteRange(toSlice(keyRange.begin), toSlice(keyRange.end));
 	}
 
-	Future<Void> commit(bool) override {
+	Future<Void> commit(bool b) override {
+		// TODO: We probably should check if there is anything to commit to SQLite.
```

Comment:
No need to check, there always will be.

Reply:
Are you sure? I added the `writeBatch == nullptr` check because I saw some crashes from trying to commit empty batches. This is probably very small either way.

Reply:
I was sure, but it turns out I wasn't quite right.

There is nothing significant that would be saved by checking and avoiding the commit, but after looking at the code it turns out that yes technically empty commit batches are possible.

The way the storage server knows which version a particular storage engine is durable to is it reads a special key in the local system space which is written in each storage server commit, whether the batch is otherwise empty or not.  The version key is updated even if the version batch contained no changes for this storage server because otherwise that version cannot be popped from the log system.

There can't be user space changes from log mutations without updating the durable version in the local system space.  I think there also can't be user space changes from data distribution without updating a different part of the local system space to indicate what key range is now resident or no longer resident.

By "local system space" I am referring to `\xff\xff` and beyond, which is unique per storage engine instance.

https://github.com/apple/foundationdb/blob/master/fdbserver/storageserver.actor.cpp#L2627

Reply:
Interesting. Thanks! I've removed the comment.